diff --git i/cbits/sqlite3.c w/cbits/sqlite3.c
index 73c69ef..528d3bd 100644
--- i/cbits/sqlite3.c
+++ w/cbits/sqlite3.c
@@ -11999,7 +11999,7 @@ struct fts5_api {
 ** The maximum value of a ?nnn wildcard that the parser will accept.
 */
 #ifndef SQLITE_MAX_VARIABLE_NUMBER
-# define SQLITE_MAX_VARIABLE_NUMBER 999
+# define SQLITE_MAX_VARIABLE_NUMBER 9999
 #endif
 
 /* Maximum page size.  The upper bound on this value is 65536.  This a limit
