diff --git a/compiler/typecheck/TcHsSyn.hs b/compiler/typecheck/TcHsSyn.hs
index 8d097f52fb..0cf5d98d44 100644
--- a/compiler/typecheck/TcHsSyn.hs
+++ b/compiler/typecheck/TcHsSyn.hs
@@ -613,7 +613,7 @@ zonkLExprs env exprs = mapM (zonkLExpr env) exprs
 zonkLExpr  env expr  = wrapLocM (zonkExpr env) expr
 
 zonkExpr env (HsVar (L l id))
-  = ASSERT2( isNothing (isDataConId_maybe id), ppr id )
+  = -- ASSERT2( isNothing (isDataConId_maybe id), ppr id )
     return (HsVar (L l (zonkIdOcc env id)))
 
 zonkExpr _ e@(HsConLikeOut {}) = return e
diff --git a/compiler/utils/Outputable.hs b/compiler/utils/Outputable.hs
index 3050fa1cf1..f31153643a 100644
--- a/compiler/utils/Outputable.hs
+++ b/compiler/utils/Outputable.hs
@@ -1202,10 +1202,10 @@ warnPprTrace True   file  line  msg x
 -- | Panic with an assertation failure, recording the given file and
 -- line number. Should typically be accessed with the ASSERT family of macros
 assertPprPanic :: HasCallStack => String -> Int -> SDoc -> a
-assertPprPanic _file _line msg
+assertPprPanic file line msg
   = pprPanic "ASSERT failed!" doc
   where
-    doc = sep [ msg, callStackDoc ]
+    doc = sep [ text"file", text file <> comma, text "line", int line, msg, callStackDoc ]
 
 pprDebugAndThen :: DynFlags -> (String -> a) -> SDoc -> SDoc -> a
 pprDebugAndThen dflags cont heading pretty_msg
