diff --git a/configure.ac b/configure.ac
index 8bed181915..431b467080 100644
--- a/configure.ac
+++ b/configure.ac
@@ -418,6 +418,10 @@ then
     AC_PATH_PROG([DllWrap],[dllwrap])
     AC_PATH_PROG([Windres],[windres])
     AC_PATH_PROG([Genlib],[genlib])
+else
+    AC_CHECK_TARGET_TOOL([Windres],[windres])
+    AC_CHECK_TARGET_TOOL([DllWrap],[dllwrap])
+    AC_CHECK_TARGET_TOOL([OBJDUMP],[objdump])
 fi
 
 DllWrapCmd="$DllWrap"
diff --git a/libraries/base/cbits/Win32Utils.c b/libraries/base/cbits/Win32Utils.c
index ce7ce97fc0..b33db04c35 100644
--- a/libraries/base/cbits/Win32Utils.c
+++ b/libraries/base/cbits/Win32Utils.c
@@ -9,6 +9,8 @@
 #include "HsBase.h"
 #include <stdbool.h>
 #include <stdint.h>
+/* Using Secure APIs */
+#define MINGW_HAS_SECURE_API 1
 #include <wchar.h>
 #include <windows.h>
 
diff --git a/rts/PathUtils.c b/rts/PathUtils.c
index 1b0b729f07..def3f7e329 100644
--- a/rts/PathUtils.c
+++ b/rts/PathUtils.c
@@ -7,6 +7,11 @@
 
 #include <libgen.h>
 #include <ctype.h>
+#if defined(mingw32_HOST_OS)
+/* Using Secure APIs */
+#define MINGW_HAS_SECURE_API 1
+#include <wchar.h>
+#endif
 
 pathchar* pathdup(pathchar *path)
 {
diff --git a/rts/win32/OSMem.c b/rts/win32/OSMem.c
index c67b95be82..534cd15fa6 100644
--- a/rts/win32/OSMem.c
+++ b/rts/win32/OSMem.c
@@ -458,7 +458,7 @@ void *osReserveHeapMemory (void *startAddress, W_ *len)
             sysErrorBelch(
                 "osReserveHeapMemory: VirtualAlloc MEM_RESERVE %llu bytes \
                 at address %p bytes failed",
-                len + MBLOCK_SIZE, startAddress);
+                *len + MBLOCK_SIZE, startAddress);
         }
         stg_exit(EXIT_FAILURE);
     }
